File: VignetteAndroid/app/src/main/kotlin/com/entativa/vignette/viewmodel/VignetteAuthViewModel.kt
package com.entativa.vignette.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.entativa.vignette.network.VignetteAuthAPIClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.regex.Pattern

/**
 * Authentication ViewModel for Vignette
 * Instagram-style authentication with enterprise-grade state management
 */
class VignetteAuthViewModel(application: Application) : AndroidViewModel(application) {
    
    private val apiClient = VignetteAuthAPIClient.getInstance(application)
    
    // MARK: - UI State
    
    data class AuthUiState(
        val isAuthenticated: Boolean = false,
        val currentUser: VignetteAuthAPIClient.User? = null,
        val isLoading: Boolean = false,
        val errorMessage: String? = null,
        val showError: Boolean = false
    )
    
    private val _uiState = MutableStateFlow(AuthUiState())
    val uiState: StateFlow<AuthUiState> = _uiState.asStateFlow()
    
    // MARK: - Sign Up Form
    
    data class SignUpFormState(
        val username: String = "",
        val email: String = "",
        val fullName: String = "",
        val password: String = "",
        
        val usernameError: String? = null,
        val emailError: String? = null,
        val fullNameError: String? = null,
        val passwordError: String? = null
    )
    
    private val _signUpFormState = MutableStateFlow(SignUpFormState())
    val signUpFormState: StateFlow<SignUpFormState> = _signUpFormState.asStateFlow()
    
    // MARK: - Login Form
    
    data class LoginFormState(
        val usernameOrEmail: String = "",
        val password: String = ""
    )
    
    private val _loginFormState = MutableStateFlow(LoginFormState())
    val loginFormState: StateFlow<LoginFormState> = _loginFormState.asStateFlow()
    
    // MARK: - Initialization
    
    init {
        checkAuthenticationStatus()
    }
    
    // MARK: - Authentication Status
    
    fun checkAuthenticationStatus() {
        viewModelScope.launch {
            if (apiClient.hasToken()) {
                apiClient.getCurrentUser().fold(
                    onSuccess = { user ->
                        _uiState.value = _uiState.value.copy(
                            isAuthenticated = true,
                            currentUser = user
                        )
                    },
                    onFailure = {
                        _uiState.value = _uiState.value.copy(
                            isAuthenticated = false,
                            currentUser = null
                        )
                    }
                )
            }
        }
    }
    
    // MARK: - Sign Up
    
    fun updateSignUpField(field: SignUpField, value: String) {
        _signUpFormState.value = when (field) {
            SignUpField.USERNAME -> _signUpFormState.value.copy(
                username = value.lowercase(),
                usernameError = null
            )
            SignUpField.EMAIL -> _signUpFormState.value.copy(
                email = value,
                emailError = null
            )
            SignUpField.FULL_NAME -> _signUpFormState.value.copy(
                fullName = value,
                fullNameError = null
            )
            SignUpField.PASSWORD -> _signUpFormState.value.copy(
                password = value,
                passwordError = null
            )
        }
    }
    
    fun signUp() {
        if (!validateSignUpForm()) return
        
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null, showError = false)
            
            val form = _signUpFormState.value
            
            apiClient.signUp(
                username = form.username.trim().lowercase(),
                email = form.email.trim().lowercase(),
                fullName = form.fullName.trim(),
                password = form.password
            ).fold(
                onSuccess = { response ->
                    _uiState.value = _uiState.value.copy(
                        isAuthenticated = true,
                        currentUser = response.data?.user,
                        isLoading = false
                    )
                    clearSignUpForm()
                },
                onFailure = { error ->
                    _uiState.value = _uiState.value.copy(
                        errorMessage = error.message ?: "Sign up failed",
                        showError = true,
                        isLoading = false
                    )
                }
            )
        }
    }
    
    private fun validateSignUpForm(): Boolean {
        val form = _signUpFormState.value
        var isValid = true
        var errors = SignUpFormState()
        
        // Validate username (Instagram-style rules)
        val trimmedUsername = form.username.trim().lowercase()
        when {
            trimmedUsername.isEmpty() -> {
                errors = errors.copy(usernameError = "Username is required")
                isValid = false
            }
            trimmedUsername.length < 3 -> {
                errors = errors.copy(usernameError = "Username must be at least 3 characters")
                isValid = false
            }
            trimmedUsername.length > 30 -> {
                errors = errors.copy(usernameError = "Username must be 30 characters or less")
                isValid = false
            }
            !isValidUsername(trimmedUsername) -> {
                errors = errors.copy(usernameError = "Username can only contain letters, numbers, periods, and underscores")
                isValid = false
            }
            trimmedUsername.startsWith(".") || trimmedUsername.endsWith(".") -> {
                errors = errors.copy(usernameError = "Username cannot start or end with a period")
                isValid = false
            }
            trimmedUsername.contains("..") -> {
                errors = errors.copy(usernameError = "Username cannot have consecutive periods")
                isValid = false
            }
        }
        
        // Validate email
        val trimmedEmail = form.email.trim()
        when {
            trimmedEmail.isEmpty() -> {
                errors = errors.copy(emailError = "Email is required")
                isValid = false
            }
            !isValidEmail(trimmedEmail) -> {
                errors = errors.copy(emailError = "Please enter a valid email address")
                isValid = false
            }
        }
        
        // Validate full name
        val trimmedFullName = form.fullName.trim()
        when {
            trimmedFullName.isEmpty() -> {
                errors = errors.copy(fullNameError = "Full name is required")
                isValid = false
            }
            trimmedFullName.length < 2 -> {
                errors = errors.copy(fullNameError = "Full name must be at least 2 characters")
                isValid = false
            }
        }
        
        // Validate password
        when {
            form.password.isEmpty() -> {
                errors = errors.copy(passwordError = "Password is required")
                isValid = false
            }
            form.password.length < 8 -> {
                errors = errors.copy(passwordError = "Password must be at least 8 characters")
                isValid = false
            }
            !form.password.any { it.isUpperCase() } -> {
                errors = errors.copy(passwordError = "Password must contain at least one uppercase letter")
                isValid = false
            }
            !form.password.any { it.isLowerCase() } -> {
                errors = errors.copy(passwordError = "Password must contain at least one lowercase letter")
                isValid = false
            }
            !form.password.any { it.isDigit() } -> {
                errors = errors.copy(passwordError = "Password must contain at least one number")
                isValid = false
            }
        }
        
        if (!isValid) {
            _signUpFormState.value = _signUpFormState.value.copy(
                usernameError = errors.usernameError,
                emailError = errors.emailError,
                fullNameError = errors.fullNameError,
                passwordError = errors.passwordError
            )
        }
        
        return isValid
    }
    
    // MARK: - Login
    
    fun updateLoginField(field: LoginField, value: String) {
        _loginFormState.value = when (field) {
            LoginField.USERNAME_OR_EMAIL -> _loginFormState.value.copy(usernameOrEmail = value)
            LoginField.PASSWORD -> _loginFormState.value.copy(password = value)
        }
    }
    
    fun login() {
        val form = _loginFormState.value
        
        if (form.usernameOrEmail.trim().isEmpty()) {
            _uiState.value = _uiState.value.copy(
                errorMessage = "Please enter your username or email",
                showError = true
            )
            return
        }
        
        if (form.password.isEmpty()) {
            _uiState.value = _uiState.value.copy(
                errorMessage = "Please enter your password",
                showError = true
            )
            return
        }
        
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null, showError = false)
            
            apiClient.login(
                usernameOrEmail = form.usernameOrEmail.trim(),
                password = form.password
            ).fold(
                onSuccess = { response ->
                    _uiState.value = _uiState.value.copy(
                        isAuthenticated = true,
                        currentUser = response.data?.user,
                        isLoading = false
                    )
                    clearLoginForm()
                },
                onFailure = { error ->
                    _uiState.value = _uiState.value.copy(
                        errorMessage = error.message ?: "Login failed",
                        showError = true,
                        isLoading = false
                    )
                }
            )
        }
    }
    
    // MARK: - Logout
    
    fun logout() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            
            apiClient.logout()
            
            _uiState.value = _uiState.value.copy(
                isAuthenticated = false,
                currentUser = null,
                isLoading = false
            )
        }
    }
    
    // MARK: - Form Management
    
    private fun clearSignUpForm() {
        _signUpFormState.value = SignUpFormState()
    }
    
    private fun clearLoginForm() {
        _loginFormState.value = LoginFormState()
    }
    
    fun clearError() {
        _uiState.value = _uiState.value.copy(errorMessage = null, showError = false)
    }
    
    // MARK: - Validation Helpers
    
    private fun isValidUsername(username: String): Boolean {
        // Instagram-style username validation: only letters, numbers, periods, and underscores
        val usernamePattern = Pattern.compile("^[a-zA-Z0-9._]+$")
        return usernamePattern.matcher(username).matches()
    }
    
    private fun isValidEmail(email: String): Boolean {
        val emailPattern = Pattern.compile(
            "[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+"
        )
        return emailPattern.matcher(email).matches()
    }
    
    // MARK: - Enums
    
    enum class SignUpField {
        USERNAME, EMAIL, FULL_NAME, PASSWORD
    }
    
    enum class LoginField {
        USERNAME_OR_EMAIL, PASSWORD
    }
}

File: VignetteiOS/ViewModels/VignetteAuthViewModel.swift
import Foundation
import SwiftUI
import Combine
import LocalAuthentication

/// Authentication ViewModel for Vignette
@MainActor
class VignetteAuthViewModel: ObservableObject {
    // MARK: - Published Properties
    
    @Published var isAuthenticated = false
    @Published var currentUser: VignetteAuthAPIClient.AuthResponse.User?
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var showError = false
    
    // Sign up form fields
    @Published var signUpUsername = ""
    @Published var signUpEmail = ""
    @Published var signUpFullName = ""
    @Published var signUpPassword = ""
    
    // Login form fields
    @Published var loginUsernameOrEmail = ""
    @Published var loginPassword = ""
    
    // Validation
    @Published var usernameError: String?
    @Published var emailError: String?
    @Published var fullNameError: String?
    @Published var passwordError: String?
    
    // Biometric auth
    @Published var biometricAuthAvailable = false
    @Published var biometricType: LABiometryType = .none
    
    private let apiClient = VignetteAuthAPIClient.shared
    private let context = LAContext()
    private var cancellables = Set<AnyCancellable>()
    
    // MARK: - Initialization
    
    init() {
        checkBiometricAvailability()
        checkAuthenticationStatus()
    }
    
    // MARK: - Authentication Status
    
    func checkAuthenticationStatus() {
        Task {
            do {
                let user = try await apiClient.getCurrentUser()
                self.currentUser = user
                self.isAuthenticated = true
            } catch {
                self.isAuthenticated = false
                self.currentUser = nil
            }
        }
    }
    
    // MARK: - Sign Up
    
    func signUp() async {
        // Validate all fields
        guard validateSignUpForm() else {
            return
        }
        
        isLoading = true
        errorMessage = nil
        showError = false
        
        do {
            let response = try await apiClient.signUp(
                username: signUpUsername.trimmingCharacters(in: .whitespaces).lowercased(),
                email: signUpEmail.trimmingCharacters(in: .whitespaces).lowercased(),
                fullName: signUpFullName.trimmingCharacters(in: .whitespaces),
                password: signUpPassword
            )
            
            self.currentUser = response.data?.user
            self.isAuthenticated = true
            self.clearSignUpForm()
            
        } catch let error as VignetteAuthError {
            self.errorMessage = error.errorDescription
            self.showError = true
        } catch {
            self.errorMessage = "An unexpected error occurred. Please try again."
            self.showError = true
        }
        
        isLoading = false
    }
    
    func validateSignUpForm() -> Bool {
        var isValid = true
        
        // Reset errors
        usernameError = nil
        emailError = nil
        fullNameError = nil
        passwordError = nil
        
        // Validate username (Instagram-style rules)
        let trimmedUsername = signUpUsername.trimmingCharacters(in: .whitespaces).lowercased()
        if trimmedUsername.isEmpty {
            usernameError = "Username is required"
            isValid = false
        } else if trimmedUsername.count < 3 {
            usernameError = "Username must be at least 3 characters"
            isValid = false
        } else if trimmedUsername.count > 30 {
            usernameError = "Username must be 30 characters or less"
            isValid = false
        } else if !isValidUsername(trimmedUsername) {
            usernameError = "Username can only contain letters, numbers, periods, and underscores"
            isValid = false
        } else if trimmedUsername.hasPrefix(".") || trimmedUsername.hasSuffix(".") {
            usernameError = "Username cannot start or end with a period"
            isValid = false
        } else if trimmedUsername.contains("..") {
            usernameError = "Username cannot have consecutive periods"
            isValid = false
        }
        
        // Validate email
        let trimmedEmail = signUpEmail.trimmingCharacters(in: .whitespaces)
        if trimmedEmail.isEmpty {
            emailError = "Email is required"
            isValid = false
        } else if !isValidEmail(trimmedEmail) {
            emailError = "Please enter a valid email address"
            isValid = false
        }
        
        // Validate full name
        let trimmedFullName = signUpFullName.trimmingCharacters(in: .whitespaces)
        if trimmedFullName.isEmpty {
            fullNameError = "Full name is required"
            isValid = false
        } else if trimmedFullName.count < 2 {
            fullNameError = "Full name must be at least 2 characters"
            isValid = false
        }
        
        // Validate password
        if signUpPassword.isEmpty {
            passwordError = "Password is required"
            isValid = false
        } else if signUpPassword.count < 8 {
            passwordError = "Password must be at least 8 characters"
            isValid = false
        } else if !signUpPassword.contains(where: { $0.isUppercase }) {
            passwordError = "Password must contain at least one uppercase letter"
            isValid = false
        } else if !signUpPassword.contains(where: { $0.isLowercase }) {
            passwordError = "Password must contain at least one lowercase letter"
            isValid = false
        } else if !signUpPassword.contains(where: { $0.isNumber }) {
            passwordError = "Password must contain at least one number"
            isValid = false
        }
        
        return isValid
    }
    
    // MARK: - Login
    
    func login() async {
        // Validate login form
        guard validateLoginForm() else {
            return
        }
        
        isLoading = true
        errorMessage = nil
        showError = false
        
        do {
            let response = try await apiClient.login(
                usernameOrEmail: loginUsernameOrEmail.trimmingCharacters(in: .whitespaces),
                password: loginPassword
            )
            
            self.currentUser = response.data?.user
            self.isAuthenticated = true
            self.clearLoginForm()
            
        } catch let error as VignetteAuthError {
            self.errorMessage = error.errorDescription
            self.showError = true
        } catch {
            self.errorMessage = "An unexpected error occurred. Please try again."
            self.showError = true
        }
        
        isLoading = false
    }
    
    func validateLoginForm() -> Bool {
        if loginUsernameOrEmail.trimmingCharacters(in: .whitespaces).isEmpty {
            errorMessage = "Please enter your username or email"
            showError = true
            return false
        }
        
        if loginPassword.isEmpty {
            errorMessage = "Please enter your password"
            showError = true
            return false
        }
        
        return true
    }
    
    // MARK: - Biometric Authentication
    
    func checkBiometricAvailability() {
        var error: NSError?
        biometricAuthAvailable = context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error)
        biometricType = context.biometryType
    }
    
    func authenticateWithBiometrics() async -> Bool {
        let context = LAContext()
        var error: NSError?
        
        guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) else {
            return false
        }
        
        do {
            let reason = biometricType == .faceID ? "Log in with Face ID" : "Log in with Touch ID"
            let success = try await context.evaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, localizedReason: reason)
            
            if success {
                // Check if we have stored credentials
                checkAuthenticationStatus()
            }
            
            return success
        } catch {
            return false
        }
    }
    
    // MARK: - Logout
    
    func logout() async {
        isLoading = true
        
        do {
            try await apiClient.logout()
            self.isAuthenticated = false
            self.currentUser = nil
        } catch {
            print("Logout error: \(error)")
            // Still log out locally even if server request fails
            self.isAuthenticated = false
            self.currentUser = nil
        }
        
        isLoading = false
    }
    
    // MARK: - Form Management
    
    func clearSignUpForm() {
        signUpUsername = ""
        signUpEmail = ""
        signUpFullName = ""
        signUpPassword = ""
        
        usernameError = nil
        emailError = nil
        fullNameError = nil
        passwordError = nil
    }
    
    func clearLoginForm() {
        loginUsernameOrEmail = ""
        loginPassword = ""
    }
    
    // MARK: - Validation Helpers
    
    func isValidUsername(_ username: String) -> Bool {
        // Instagram-style username validation
        // Only letters, numbers, periods, and underscores
        let usernameRegex = "^[a-zA-Z0-9._]+$"
        let usernamePredicate = NSPredicate(format:"SELF MATCHES %@", usernameRegex)
        return usernamePredicate.evaluate(with: username)
    }
    
    func isValidEmail(_ email: String) -> Bool {
        let emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
        let emailPredicate = NSPredicate(format:"SELF MATCHES %@", emailRegex)
        return emailPredicate.evaluate(with: email)
    }
}
